cherab/solps/formats/balance.py

```diff
@@ -0,0 +1,192 @@
+	
+    # Load SOLPS mesh geometry
+    cr_x = copy.deepcopy(fhandle.variables['crx'].data)
+    cr_z = copy.deepcopy(fhandle.variables['cry'].data)
+    vol = copy.deepcopy(fhandle.variables['vol'].data)
+	
+    # Re-arrange the array dimensions in the way CHERAB expects...
+    cr_x = swapaxes(cr_x,0,2)
+    cr_x = swapaxes(cr_x,0,1)
+    cr_z = swapaxes(cr_z,0,2)
+    cr_z = swapaxes(cr_z,0,1)
+	
+    # Create the SOLPS mesh
+    mesh = SOLPSMesh(cr_x,cr_z,vol)	
+
+    sim = SOLPSSimulation(mesh)
+    ni = mesh.nx
+    nj = mesh.ny
+
+    # TODO: add code to load SOLPS velocities and magnetic field from files
+
+    # Load electron species
+    sim._electron_temperature = copy.deepcopy(fhandle.variables['te'].data)/Q
+    sim._electron_density = copy.deepcopy(fhandle.variables['ne'].data)
+
+    ##########################################
+    # Load each plasma species in simulation #
+    ##########################################
+
+    sim._species_list = []
+    n_species = len(fhandle.variables['am'].data)
+	
+    for i in range(n_species):
+
+        # Extract the nuclear charge	
+        if fhandle.variables['species'].data[i,1] == b'D':
+	        zn = 1
+        if fhandle.variables['species'].data[i,1] == b'C':
+            zn = 6
+        if fhandle.variables['species'].data[i,1] == b'N':
+            zn = 7
+        if fhandle.variables['species'].data[i,1] == b'N' and fhandle.variables['species'].data[i,2] == b'e':
+            zn = 10
+        if fhandle.variables['species'].data[i,1] == b'A' and fhandle.variables['species'].data[i,2] == b'r':
+            zn = 18
+	
+        am = float(fhandle.variables['am'].data[i])  # Atomic mass
+        charge = int(fhandle.variables['za'].data[i])  # Ionisation/charge
+        species = _popular_species[(zn, am)]
+		
+        # If we only need to populate species_list, there is probably a faster way to do this...		
+        sim.species_list.append(species.symbol + str(charge))
+
+    tmp = copy.deepcopy(fhandle.variables['na'].data)
+    tmp = swapaxes(tmp,0,2)
+    tmp = swapaxes(tmp,0,1)
+    sim._species_density = tmp
+
+    # Make Mesh Interpolator function for inside/outside mesh test.
+    inside_outside_data = ones(mesh.num_tris)
+    inside_outside = AxisymmetricMapper(Discrete2DMesh(mesh.vertex_coords, mesh.triangles, inside_outside_data, limit=False))
+    sim._inside_mesh = inside_outside
+	
+    # Load the neutrals data
+    if 'D0' in sim.species_list:
+        for i in arange(len(sim.species_list)):
+            if sim.species_list[i] == 'D0':
+                D0_indx = i
+
+    # Replace the deuterium neutrals density (from the fluid neutrals model by default) with
+    # the values calculated by EIRENE - do the same for other neutrals?
+    if 'dab2' in fhandle.variables.keys():
+        sim.species_density[:,:,D0_indx] = copy.deepcopy(fhandle.variables['dab2'].data[0,:,0:-2])
+        eirene_run = True
+    else:
+        eirene_run = False
```

Comment:
There are a few issues here:
1. `D0_indx` is only defined if D0 appears in `sim.species_list`. If the balance file somehow has a `dab2` variable but no D0 species then `D0_indx` in line 134 will be undefined and you'll get an `UnboundLocalError` exception raised. For robustness, you could set `D0_indx` to a sentinel value such as `None` beforehand, and only update the neutral density if `D0_indx` is not `None`.
2. Lines 125-129 can be rewritten as: `D0_indx = sim.species_list.index('D0')`. Python has batteries included!
3. Same comment about deepcopy.

You can re-write this section more compactly as the following:
```python
try:
    D0_indx = sim.species_list.index('D0')
except ValueError:  # No 'D0'
    eirene_run = False
    # Could instead raise an error here, if a lack of D0 is a critical error.
else:
    if 'dab2' in fhandle.variables.keys():
        sim.species_density[:, :, D0_indx] = fhandle.variables['dab2'].data[0, :, 0:-2].copy()
        eirene_run = True
    else:
        eirene_run = False
```

I also ran into an error in this part when testing against a balance.nc file I have: 
```python
In [59]: sim = load_solps_from_balance('/home/jlovell/Bolometer/SXD phantoms/Full data/seed_8.nc')       
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-59-de7e0aeae2c8> in <module>
----> 1 sim = load_solps_from_balance('/home/jlovell/Bolometer/SXD phantoms/Full data/seed_8.nc')

~/cherab/cherab_solps/cherab/solps/formats/balance.py in load_solps_from_balance(balance_filename, debug)
    132     # the values calculated by EIRENE - do the same for other neutrals?
    133     if 'dab2' in fhandle.variables.keys():
--> 134         sim.species_density[:,:,D0_indx] = copy.deepcopy(fhandle.variables['dab2'].data[0,:,0:-2])
    135         eirene_run = True
    136     else:

ValueError: could not broadcast input array from shape (34,151) into shape (34,148)
```
Looks like there is a dimension issue happening somewhere.